Develop.property list

illiapelypenko opened the pull request:
Add property-list page
Fix code-review issues

illiapelypenko committed:
Fix problems with state managment

illiapelypenko committed:
Add search tip
Handle empty search submitting

illiapelypenko committed:
Implement base styling of property list

illiapelypenko committed:
Add load more button logic

illiapelypenko committed:
Create page for property list
Delete sass rule from webpack
Add constants for dom-selectors
Api key set as constant
Change event listeners name
Change spinner implementation
Add public-class-fields babel polifill

illiapelypenko committed:
Add results-buffer to property search
Add debounce for suggestion fetching

illiapelypenko committed:
Fix main components' width

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,54 @@
 import { state } from '../index';
 
+const KEY = '897733c2d3msh13b53444c4f53c4p152d7fjsn04123a994089';
```
API_KEY

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,54 @@
-      encodeURI(
-        `https://realtor.p.rapidapi.com/locations/auto-complete?input=${location}`
-      ),
+      encodeURI(`https://realtor.p.rapidapi.com/locations/auto-complete?input=${location}`),
```
Could you please create variables for URL and Queries and other constant values, for instance:
BASE_URL = 'https://realtor.p.rapidapi.com/locations/'
AUTOCOMPLETE_URL=BASE_URL+ 'auto-complete'

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,54 @@
+export async function getProperties() {
+  const { city, state_code } = state.currentSearchListItem;
+  const mapKey = JSON.stringify({ city, state_code });
+  if (propertiesBuffer.has(mapKey)) return (state.properties = propertiesBuffer.get(mapKey));
```
Sometimes I write in one line, but frankly speaking:
if (propertiesBuffer.has(mapKey)) {
   return state.properties = propertiesBuffer.get(mapKey)
}
, better to read

uladzislau-stuk commented on src/components/eventListeners.js:
```diff
@@ -1,68 +1,133 @@
-  RECENT_SEARCHES_LIST.addEventListener('click', onRecentSearchClick);
+  searchTextInput.addEventListener('focus', onSearchFocus, true);
+  searchTextInput.addEventListener('blur', onSearchBlur, true);
+  searchTextInput.addEventListener('input', onSearchInput);
```
I guess you don't need anonymous self-invoke function, just write:
searchTextInput.addEventListener('focus', onSearchFocus(), true)
, and remove anonymous self-invoke function

uladzislau-stuk commented on src/components/eventListeners.js:
```diff
@@ -1,68 +1,133 @@
-    state.suggestions[Math.floor(Math.random() * state.suggestions.length)];
-  SEARCH_TEXT_INPUT.value = `${state.currentSearchItem.area_type}, ${state.currentSearchItem.city}, ${state.currentSearchItem.state_code}`;
+  state.currentSearchItem = state.suggestions[Math.floor(Math.random() * state.suggestions.length)];
+  searchTextInput.value = `${state.currentSearchItem.area_type}, ${state.currentSearchItem.city}, ${state.currentSearchItem.state_code}`;
```
use:
const { area_type, city, state_code } = state.currentSearchItem;
searchTextInput.value = `${area_type}, ${city}, ${ state_code}`,

I would be better to create getter:
get<Yourname>Value() {
  const { area_type, city, state_code } = state.currentSearchItem;
  return `${area_type}, ${city}, ${ state_code}`,
}

function onSuggestionClick() {
 //...
  searchTextInput.value = get<Yourname>Value()
}

uladzislau-stuk commented on src/components/view.js:
```diff
@@ -1,35 +1,98 @@
+  const suggestionToShow = 4;
+  const suggestionHeight = 56;
+
+  while (suggestionList.firstChild) {
+    suggestionList.removeChild(suggestionList.firstChild);
   }
```
Please move this to separate function, you can create file utils and import helpers from there

uladzislau-stuk commented on src/components/view.js:
```diff
@@ -1,35 +1,98 @@
+  propertyList,
+  matchesCount,
+  loadMoreButton,
+} from './domConsts';
```
rename please domConsts.js to elements.js

uladzislau-stuk commented on src/components/view.js:
```diff
@@ -1,35 +1,98 @@
+  const toRender = 20;
+  let properties = state.properties.slice(state.propertyOffset, state.propertyOffset + toRender);
+  matchesCount.innerHTML = `${state.propertyOffset + properties.length} of ${
+    state.properties.length
+  } matches`;
+
+  for (let property of properties) {
+    const {
+      address: { city, line },
+      price,
+      thumbnail,
+    } = property;
+
+    if (!thumbnail) continue;
+
+    const li = document.createElement('li');
+    li.classList.add('property-item');
+
+    const img = document.createElement('img');
+    img.src = thumbnail;
+    img.alt = 'thumbnail';
+    img.classList.add('property-item__thumbnail');
+
+    const div = document.createElement('div');
+    div.classList.add('property-item__info');
+
+    const priceElement = document.createElement('p');
+    priceElement.classList.add('property-item__price');
+    priceElement.innerHTML = `$ ${price}`;
+
+    const locationElement = document.createElement('p');
+    locationElement.classList.add('property-item__location');
+    locationElement.innerHTML = `${city}, ${line}`;
+
+    li.appendChild(img);
+    li.appendChild(div);
+    div.appendChild(priceElement);
+    div.appendChild(locationElement);
+
+    propertyList.appendChild(li);
```
If function more than 10-15 lines it is sign that you did something wrong, please think about refactoring such functions

uladzislau-stuk commented:
Could you please try to avoid prefixing branches with the name it parent branch, property-list - ok

illiapelypenko committed:
Create utils functions
Rename file domConsts to elements
Split createPropertyItemFunction into two functions
Create string consts in api functions

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,62 @@
+
   try {
+    const BASE_URL = 'https://realtor.p.rapidapi.com';
+    const LOCATIONS_AUTOCOMPLETE_URL = BASE_URL + '/locations/auto-complete';
```
Move to to the top of the file

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,62 @@
+
+  try {
+    const BASE_URL = 'https://realtor.p.rapidapi.com';
+    const PROPERTIES_FOR_SALE_URL = BASE_URL + '/properties/v2/list-for-sale';
```
Move to the top

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,62 @@
+    const res = await fetch(
+      `${BASE_URL + PROPERTIES_FOR_SALE_URL}?sort=relevance&city=${encodeURI(
+        city
+      )}&limit=500&offset=0&state_code=${encodeURI(state_code)}`,
```
const urLParams = `sort=relevance&city=${city}&limit=500&offset=0&state_code=${state_code}`
PROPERTIES_FOR_SALE_URL + encodeUrl(urLParams)

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,62 @@
+      {
+        method: 'GET',
+        headers: {
+          'x-rapidapi-host': 'realtor.p.rapidapi.com',
```
Use constant

uladzislau-stuk commented on src/components/eventListeners.js:
```diff
@@ -1,68 +1,131 @@
-    item => item._id === e.target.dataset.key
+  if (e.target.nodeName !== 'LI') return;
+
+  const index = Array.from(suggestionList.children).findIndex(
+    item => item.outerText === e.target.innerText
```
Create util
const index = Array.from(suggestionList.children).findIndex(
    item => item.outerText === e.target.innerText
  );

uladzislau-stuk commented on src/components/utils.js:
```diff
@@ -0,0 +1,10 @@
+export function getSearchItemInfo() {
+  const { area_type, city, state_code } = state.currentSearchItem;
+  return `${area_type}, ${city}, ${state_code}`;
+}
```
Pass state as param, don't use it globally
function getSearchItemInfo(state) {
  const { area_type, city, state_code } = state.currentSearchItem;
  return `${area_type}, ${city}, ${state_code}`;
}

uladzislau-stuk commented on src/components/view.js:
```diff
@@ -1,35 +1,99 @@
+}
+
+export function renderProperties() {
+  if (state.properties.length - state.propertyOffset > 20) loadMoreButton.style.display = 'block';
+  else loadMoreButton.style.display = 'none';
```
loadMoreButton.style.display = state.properties.length - state.propertyOffset > 20 ? 'block' : 'none'

illiapelypenko committed:
Refactor getSearchItemInfo func
Create getNodeElementIndexFromNodeList func
Make some string params a constats

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,63 @@
-        },
-      }
-    );
+    const URL_PARAMS = encodeURI(`?input=${location}`);
```
just urlParams

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,63 @@
+    const fetchMethod = 'GET';
+    const xRapidapiHost = 'realtor.p.rapidapi.com';
+
+    const res = await fetch(BASE_URL + LOCATIONS_AUTOCOMPLETE_URL + URL_PARAMS, {
```
const BASE_URL = 'https://realtor.p.rapidapi.com';
const LOCATIONS_AUTOCOMPLETE_URL = BASE_URL + '/locations/auto-complete';
BASE_URL + LOCATIONS_AUTOCOMPLETE_URL = 'https://realtor.p.rapidapi.com' + 'https://realtor.p.rapidapi.com' + '/locations/auto-complete'

Are you sure it's correct url?

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,63 @@
-    );
+    const URL_PARAMS = encodeURI(`?input=${location}`);
+    const fetchMethod = 'GET';
+    const xRapidapiHost = 'realtor.p.rapidapi.com';
```
Please move to constant

uladzislau-stuk commented on src/components/api.js:
```diff
@@ -1,50 +1,63 @@
-      }
-    );
+    const URL_PARAMS = encodeURI(`?input=${location}`);
+    const fetchMethod = 'GET';
```
fetch use 'GET' by default

uladzislau-stuk commented on src/components/utils.js:
```diff
@@ -0,0 +1,14 @@
+}
+
+export function getNodeElementIndexFromNodeList(nodeElement, nodeList) {
+  return Array.from(nodeList.children).findIndex(item => item.outerText === nodeElement.innerText);
```
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/outerText

uladzislau-stuk reviewed (changes requested):
Could you please check application before sending on review. Fix errors.

![image](https://user-images.githubusercontent.com/43731223/91443392-f3f89f00-e87b-11ea-95d7-e4e0e6412ae6.png)

Commit:
Fix api urls
Improve props filter method
